src/ddmd/backend/cgelem.c

```diff
@@ -1900,20 +1900,65 @@ STATIC elem * elcond(elem *e, goal_t goal)
             elem *ec2 = e->E2->E2;
 
             if (tyintegral(ty) && ec1->Eoper == OPconst && ec2->Eoper == OPconst)
-            {   targ_llong i1,i2;
-                targ_llong b;
-
-                i1 = el_tolong(ec1);
-                i2 = el_tolong(ec2);
+            {
+                targ_llong i1 = el_tolong(ec1);
+                targ_llong i2 = el_tolong(ec2);
+                tym_t ty1 = tybasic(e1->Ety);
 
-                /* If b is an integer with only 1 bit set then          */
-                /*   replace ((a & b) ? b : 0) with (a & b)             */
-                /*   replace ((a & b) ? 0 : b) with ((a & b) ^ b)       */
-                if (e1->Eoper == OPand && e1->E2->Eoper == OPconst &&
+                if ((ty1 == TYbool && !OTlogical(e1->Eoper) || e1->Eoper == OPand && e1->E2->Eoper == OPconst) &&
                     tysize(ty) == tysize(ec1->Ety))
                 {
-                    b = el_tolong(e1->E2);
-                    if (ispow2(b) != -1)        /* if only 1 bit is set */
+                    targ_llong b = ty1 == TYbool ? 1 : el_tolong(e1->E2);
+
+                    if (b == 1 && ispow2(i1 - i2))
```

Comment:
Also, why is this checking `ispow2` is non-zero, while below it checks `ispow2(b) != -1`?